Update `azurerm_backup_policy_vm` -  check for <7 daily backups

akonrath opened the pull request:
Fixes #7792 

(If there's a better way to do this, please let me know.)

```
$ make acctests SERVICE='recoveryservices' TESTARGS='-run=TestAccAzureRMBackupProtectionPolicyVM' TESTTIMEOUT='60m'

$ make website
$ visit http://localhost:4567/docs/providers/azurerm/r/backup_policy_vm.html
```

mbfrahry commented on website/docs/r/backup_policy_vm.markdown:
```diff
@@ -99,7 +99,7 @@ The `backup` block supports:
 The `retention_daily` block supports:
 
-* `count` - (Required) The number of daily backups to keep. Must be between `1` and `9999`
+* `count` - (Required) The number of daily backups to keep. Must be between `1` and `9999`. This can no longer be created or updated with a value less than `7`.
```
Only older users should experience this issue so we can change the documentation to reflect that we only support `7` to `9999` so we don't confuse newer users trying to set this up

```suggestion
* `count` - (Required) The number of daily backups to keep. Must be between `7` and `9999`.
```

mbfrahry commented on azurerm/internal/services/recoveryservices/resource_arm_backup_policy_vm.go:
```diff
@@ -263,6 +264,11 @@ func resourceArmBackupProtectionPolicyVM() *schema.Resource {
 				}
 
+				// Less than 7 daily backups is no longer supported for create/update
+				if diff.Get("retention_daily.0.count").(int) < 7 {
```
I wish we could set this up for users who are still using setups that are less than 7 but we have to make a best effort here. I don't think this section will work because this code will trigger regardless of an update or not and this will break someone who is using the old system. 

I think our best effort is to just prevent users from creating this resource with less than a count of 7. It'll be a little more work and won't prevent someone from trying to update `count` to less than 7 but it'll make sure we don't break users that are currently using less than 7. 

That change is going to require a bit more work though. We'll want to update this [method](https://github.com/terraform-providers/terraform-provider-azurerm/blob/f0c4b1f503dba0b0ababd33d7626f218327a38d7/azurerm/internal/services/recoveryservices/resource_arm_backup_policy_vm.go#L487) by using `d.IsNewResource()` and we error out if it is a new resource with a count less than 7.

Let me know if that makes sense and if not, I can go into more detail.

akonrath commented:
Hi @mbfrahry ! Thank  you so much for taking the time to review this.

I think your suggestion regarding the documentation is a good idea, and will help set the stage for the next major release. It is also less confusing to others. I have committed your suggestion.

Regarding the `count` change, yes, the main goal is to still support people that have already set a `count` of less than 7. The logic of the commit I was targeting was something like:
- if `count` has changed to an integer less than 7 for create/update, fail. (prevent people from creating/updating to less than 7)
- if `count` has not changed and is less than 7, pass. (retain compatibility)

Is there anyway to capture that logic for a specific input such as `count`? My only worry is that if someone that is "grandfathered in" to less than 7 backups tries to change to less than 7, it'll still throw the confusing error if we just target new resource creation.  They'd be able to view the (updated) docs but still might be confusing.

akonrath commented:
I've updated the PR with my goals listed above (the original goals I was aiming for). I moved this out of the initial diff (which breaks compatibility) and moved it to test with `d.HasChange` for the daily count, and perform a comparison if less than 7.

I tested this creating with a daily count of 7, re-compiling with the integer set to 8, and attempting to update (it fails with the error message).

tombuildsstuff commented on azurerm/internal/services/recoveryservices/resource_arm_backup_policy_vm.go:
```diff
@@ -313,6 +314,10 @@ func resourceArmBackupProtectionPolicyVMCreateUpdate(d *schema.ResourceData, met
 		}
 	}
+	// Less than 7 daily backups is no longer supported for create/update
+	if d.HasChange("retention_daily.0.count") && d.Get("retention_daily.0.count").(int) < 7 {
```
where existing resources are grandfathered in (and there's been multiple years worth, so this isn't likely going to change anytime soon) - I think this wants to be:

```suggestion
	if (d.IsNewResource() || d.HasChange("retention_daily.0.count")) && d.Get("retention_daily.0.count").(int) < 7) {
```

which'll handle:

- new resources must have at least 7 days
- existing resources (with less than 7 days) can stay that way providing they're not updated

tombuildsstuff commented on azurerm/internal/services/recoveryservices/resource_arm_backup_policy_vm.go:
```diff
@@ -313,6 +314,10 @@ func resourceArmBackupProtectionPolicyVMCreateUpdate(d *schema.ResourceData, met
 	}
+	// Less than 7 daily backups is no longer supported for create/update
+	if d.HasChange("retention_daily.0.count") && d.Get("retention_daily.0.count").(int) < 7 {
+		return fmt.Errorf("creating/updating `retention_daily` count of less than 7 is not supported")
```
in this instance however we probably want to clarify what's going on here:

```suggestion
		return fmt.Errorf("The Azure API has recently changed behaviour so that provisioning a `count` for the `retention_daily` field can no longer be less than 7 days for new/updates to existing Backup Policies. Please ensure that `count` is less than 7, currently %d", d.Get("retention_daily.0.count").(int))
```

tombuildsstuff commented on website/docs/r/backup_policy_vm.markdown:
```diff
@@ -99,7 +99,7 @@ The `backup` block supports:
 The `retention_daily` block supports:
 
-* `count` - (Required) The number of daily backups to keep. Must be between `1` and `9999`
+* `count` - (Required) The number of daily backups to keep. Must be between `7` and `9999`.
```
might be worth clarifying that older resources can continue to exist:

```suggestion
* `count` - (Required) The number of daily backups to keep. Must be between `7` and `9999`.

~> **Note:** Azure previously allows this field to be set to a minimum of 1 day - but for new resources/to update this value on existing Backup Policies, this value must be at least 7 days.
```

tombuildsstuff replied:
might be worth clarifying that older resources can continue to exist:

```suggestion
* `count` - (Required) The number of daily backups to keep. Must be between `7` and `9999`.

~> **Note:** Azure previously allows this field to be set to a minimum of 1 (day) - but for new resources/to update this value on existing Backup Policies - this value must now be at least 7 (days).
```

tombuildsstuff reviewed (changes requested):
hey @akonrath 

Thanks for pushing those changes - I've taken a look through and left some comments inline but if we can fix those up then this otherwise LGTM 👍 

Thanks!

akonrath commented:
@tombuildsstuff  - Thank you so much for helping; you rock!

1. That conditional you've suggested looks to be what we're aiming for. I've updated the second half of it to cover a range of 1 to 7 because I found during testing (thanks to your increased output verbosity), that going from daily to weekly registers the count to 0, which breaks the acceptance tests, and prevents that change from taking place.

2. Great idea about enhancing the verbosity on the docs. This will definitely give more context to people. I've merged the commit.

Acceptance tests are now passing, and I also tested outside of them:
```
$ make acctests SERVICE='recoveryservices' TESTARGS='-run=TestAccAzureRMBackupProtectionPolicyVM' TESTTIMEOUT='60m'
=> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/recoveryservices/tests/ -run=TestAccAzureRMBackupProtectionPolicyVM -timeout 60m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_basicDaily
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_basicDaily
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_requiresImport
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_requiresImport
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_basicWeekly
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_basicWeekly
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_completeDaily
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_completeDaily
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_completeWeekly
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_completeWeekly
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_updateDaily
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_updateDaily
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_updateWeekly
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_updateWeekly
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_updateDailyToWeekly
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_updateDailyToWeekly
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_updateWeeklyToDaily
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_updateWeeklyToDaily
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_updateWeeklyToPartial
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_updateWeeklyToPartial
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_basicDaily
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_updateWeekly
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_basicWeekly
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_completeDaily
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_updateDaily
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_updateWeeklyToPartial
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_completeWeekly
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_requiresImport
--- PASS: TestAccAzureRMBackupProtectionPolicyVM_basicWeekly (104.87s)
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_updateWeeklyToDaily
--- PASS: TestAccAzureRMBackupProtectionPolicyVM_basicDaily (104.87s)
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_updateDailyToWeekly
--- PASS: TestAccAzureRMBackupProtectionPolicyVM_completeDaily (108.41s)
--- PASS: TestAccAzureRMBackupProtectionPolicyVM_requiresImport (113.14s)
--- PASS: TestAccAzureRMBackupProtectionPolicyVM_completeWeekly (166.61s)
--- PASS: TestAccAzureRMBackupProtectionPolicyVM_updateDaily (190.86s)
--- PASS: TestAccAzureRMBackupProtectionPolicyVM_updateWeeklyToPartial (190.86s)
--- PASS: TestAccAzureRMBackupProtectionPolicyVM_updateWeekly (212.58s)
--- PASS: TestAccAzureRMBackupProtectionPolicyVM_updateDailyToWeekly (158.25s)
--- PASS: TestAccAzureRMBackupProtectionPolicyVM_updateWeeklyToDaily (158.26s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/recoveryservices/tests	263.165s
```

katbyte reviewed (approved):
Thanks @akonrath - LGTM 👍
